image/manifest.go

```diff
@@ -240,3 +240,38 @@ loop:
 	}
 	return nil
 }
+
+// ValidateManifestMediaType validate the manifest schema media-type
+// and its descendants fields media type, such as config and layers,
+// to make sure they match to spec definition, or returns an error if
+// the validation failed.
+func ValidateManifestMediaType(r io.ReadCloser) error {
+	header := v1.Manifest{}
+
+	buf, err := ioutil.ReadAll(r)
+	if err != nil {
+		return errors.Wrapf(err, "error reading the io stream")
+	}
+
+	err = r.Close()
+	if err != nil {
+		return errors.Wrapf(err, "error close the io stream")
+	}
+
+	err = json.NewDecoder(bytes.NewReader(buf)).Decode(&header)
+	if err != nil {
+		return errors.Wrap(err, "manifest format mismatch")
+	}
+
+	if header.Config.MediaType != string(v1.MediaTypeImageConfig) {
+		return fmt.Errorf("illegal config mediaType: %s", header.Config.MediaType)
+	}
+
+	for _, layer := range header.Layers {
+		if layer.MediaType != string(v1.MediaTypeImageLayer) {
+			return fmt.Errorf("illegal layer mediaType: %s", layer.MediaType)
+		}
+	}
+
+	return schema.MediaTypeManifest.Validate(bytes.NewReader(buf))
```

Comment:
I don't see how this code is checking descendants. It decodes the manifest, then checks the mediatypes of the layers.

It looks like there is a bug in the validation. Go fix that rather than just adding more code.